tests/Integration/Behaviour/Features/Context/OrderFeatureContext.php

```diff
@@ -82,6 +116,382 @@ public function validateCartWithPaymentModule($paymentModuleName)
         $kernel = $previousKernel;
     }
 
+    /**
+     * @When /^(\d+) items? of product "(.+)" are added in my cart order, with prices (\d+\.\d+) tax excluded and (\d+\.\d+) tax included$/
+     */
+    public function addProductInCartOrder($quantity, $productName, $priceTaxExcl, $priceTaxIncl)
+    {
+        $cart = $this->getCurrentCart();
+        $order = $this->getCurrentCartOrder();
+        $this->productFeatureContext->checkProductWithNameExists($productName);
+        $product = $this->productFeatureContext->getProductWithName($productName);
+        $this->addProductToOrder($order, [
+            'product_id' => $product->id,
+            'product_price_tax_excl' => $priceTaxExcl,
+            'product_price_tax_incl' => $priceTaxIncl,
+            'product_quantity' => $quantity,
+        ]);
+        // restore correct cart since previous method has overridden it
+        Context::getContext()->cart = $cart;
+    }
+
+    /**
+     * Duplicate from AdminOrderController::addProductToOrder
+     *
+     * @param Order $order
+     * @param array $product_informations
+     * @param array $invoice_informations
+     * @param bool $warehouseId
+     *
+     * @return array
+     */
+    protected function addProductToOrder(Order $order, array $product_informations, array $invoice_informations = [], $warehouseId = false)
```

Comment:
What a huge function, any possibility to split it?

Reply:
Actually it's a duplicate from the method in controllers/admin/AdminOrdersController.php but as this method is protected, we can't use it. So @matks has adviced me to duplicate it.

Reply:
Mmmm, the problem is that if we need to fix this again we will have to maintain both the controller AND the behat context.. So it would be better if this can be extracted in a separate class/service